tests/integration/test_charm.py

```diff
@@ -0,0 +1,186 @@
+    LEGEND_DB_APP_NAME,
+    SDLC_APP_NAME,
+    ENGINE_APP_NAME,
+    STUDIO_APP_NAME,
+    GITLAB_INTEGRATOR_APP_NAME,
+]
+OTHER_APPS = [
+    MONGODB_APP_NAME,
+    LEGEND_DB_APP_NAME,
+    ENGINE_APP_NAME,
+    STUDIO_APP_NAME,
+    GITLAB_INTEGRATOR_APP_NAME,
+]
+
+APP_PORTS = {
+    ENGINE_APP_NAME: 6060,
+    SDLC_APP_NAME: 7070,
+    STUDIO_APP_NAME: 8080,
+}
+APP_ROUTES = {
+    ENGINE_APP_NAME: "/engine",
+    SDLC_APP_NAME: "/api",
+    STUDIO_APP_NAME: "/studio",
+}
+
+
+@tenacity.retry(
+    retry=tenacity.retry_if_result(lambda x: x is False),
+    stop=tenacity.stop_after_attempt(10),
+    wait=tenacity.wait_exponential(multiplier=1, min=5, max=30),
+)
+def check_legend_connection(app_name, url, headers=None):
+    """Tests the connection to the given Legend URL.
+
+    When connecting to the FINOS Legend Applications, it will redirect to gitlab for
+    authentication, which will not be a 20X status code. We will receive 302 Found
+    instead.
+    Returns True if the response status code is 302, False in any other case.
+    """
+    logger.info("Trying to access %s...", app_name)
+    try:
+        response = requests.get(url, headers=headers, allow_redirects=False, timeout=10)
+        return response.status_code == 302
+    except Exception as ex:
+        logger.info(ex)
+    return False
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: pytest_plugin.OpsTest):
+    """Build the charm-under-test and deploy it together with related charms."""
+    # build and deploy charm from local source folder
+    charm = await ops_test.build_charm(".")
+    resources = {"sdlc-image": METADATA["resources"]["sdlc-image"]["upstream-source"]}
+    await ops_test.model.deploy(charm, resources=resources, application_name=APP_NAME)
+
+    for name in OTHER_APPS:
+        await ops_test.model.deploy(name, channel="edge")
+
+    await ops_test.model.add_relation(LEGEND_DB_APP_NAME, MONGODB_APP_NAME)
+    for app in LEGEND_APPS_CONFIG:
+        await ops_test.model.add_relation(LEGEND_DB_APP_NAME, app)
+        await ops_test.model.add_relation(GITLAB_INTEGRATOR_APP_NAME, app)
+
+    for app in LEGEND_APPS:
+        await ops_test.model.add_relation(STUDIO_APP_NAME, app)
+
+
+@pytest.mark.abort_on_fail
+async def test_config_gitlab(ops_test: pytest_plugin.OpsTest):
+    """Set the gitlab.com's application secret ID and name.
+
+    After setting the application secret ID and name, the charms should become active.
+    """
+    # We need valid information here. These should be passed in through the env variables.
+    gitlab_integrator = ops_test.model.applications["finos-legend-gitlab-integrator-k8s"]
+    await gitlab_integrator.set_config(
+        {"gitlab-client-id": GITLAB_CLIENT_ID, "gitlab-client-secret": GITLAB_CLIENT_SECRET}
+    )
+
+    # Wait for all the charms to become Active.
+    for name in APP_LIST:
+        await ops_test.model.wait_for_idle(apps=APP_LIST, status="active", timeout=2000)
```

Comment:
since we're now passing the ``APP_LIST`` to ``wait_for_idle``, we don't need to have the for in the line above. We don't have to call ``wait_for_idle`` N times.

Reply:
Solved.